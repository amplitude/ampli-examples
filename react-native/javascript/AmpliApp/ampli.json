{
  "OrgId": "132559",
  "WorkspaceId": "77b37977-cb3a-42eb-bce3-09f5f7c3adb7",
  "SourceId": "ee30d28b-4770-4651-8bc6-bfaa48dd8117",
  "Runtime": "react-native:javascript-ampli",
  "Path": "./src/ampli",
  "Branch": "main",
  "Version": "0.0.0",
  "VersionId": "79154a50-f057-4db5-9755-775e4e9f05e6"
}