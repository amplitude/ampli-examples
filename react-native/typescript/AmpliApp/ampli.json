{
  "OrgId": "132559",
  "WorkspaceId": "77b37977-cb3a-42eb-bce3-09f5f7c3adb7",
  "SourceId": "4afe79f0-c63a-484c-8589-11380377ead8",
  "Runtime": "react-native:typescript-ampli",
  "Path": "./src/ampli",
  "Branch": "main",
  "Version": "0.0.0"
}
