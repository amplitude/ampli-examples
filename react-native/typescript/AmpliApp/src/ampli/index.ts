/* tslint:disable */
/* eslint-disable eslint-comments/no-unlimited-disable */
/* eslint-disable */
/**
 * Ampli - A strong typed wrapper for your Analytics
 *
 * This file is generated by Amplitude.
 * To update run 'ampli pull react-native-typescript-ampli'
 *
 * Required dependencies: @amplitude/react-native@^2.11.0
 * Tracking Plan Version: 0
 * Build: 1.0.0
 * Runtime: react-native:typescript-ampli
 *
 * [View Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest)
 *
 * [Full Setup Instructions](https://data.amplitude.com/test-codegen/Test%20Codegen/implementation/react-native-typescript-ampli)
 */

import {
  Amplitude,
  Identify as AmplitudeIdentify,
  BaseEvent as Event,
  MiddlewareExtra,
  Plan,
} from '@amplitude/react-native';

export type EventOptions = {
  userId?: string;
  deviceId?: string;
}

export type Environment = 'development' | 'production';

export const ApiKey: Record<Environment, string> = {
  development: '',
  production: ''
};

export type ClientOptions = {
  plan?: Plan,
}

export interface LoadOptions {
  environment?: Environment;
  disabled?: boolean;
  client?: {
    apiKey?: string;
    instance?: Amplitude;
    options?: ClientOptions;
  }
}

export const DefaultOptions: ClientOptions = {
  plan: {
    branch: 'main',
    source: 'react-native-typescript-ampli',
    version: '0',
    versionId: '79154a50-f057-4db5-9755-775e4e9f05e6',
  }
}

export interface IdentifyProperties {
  /**
   * Description for identify optionalArray
   *
   * | Rule | Value |
   * |---|---|
   * | Unique Items | false |
   * | Item Type | string |
   */
  optionalArray?: string[];
  /**
   * Description for identify requiredNumber
   *
   * | Rule | Value |
   * |---|---|
   * | Type | number |
   */
  requiredNumber: number;
}

export interface GroupProperties {
  /**
   * Description for group optionalString
   */
  optionalString?: string;
  /**
   * Description for group requiredBoolean
   */
  requiredBoolean: boolean;
}

export interface EventMaxIntForTestProperties {
  /**
   * property to test schema validation
   *
   * | Rule | Value |
   * |---|---|
   * | Type | integer |
   * | Max Value | 10 |
   */
  intMax10: number;
}

export interface EventObjectTypesProperties {
  /**
   * Property Object Type
   */
  requiredObject: {
    [k: string]: any;
  };
  /**
   * Property Object Array Type
   */
  requiredObjectArray: any[];
}

export interface EventWithAllPropertiesProperties {
  /**
   * Event 2 Property - Optional String    *     * Examples:    * Some string, or another
   */
  optionalString?: string;
  /**
   * Event 2 Property - Array
   *
   * | Rule | Value |
   * |---|---|
   * | Min Items | 0 |
   * | Item Type | string |
   */
  requiredArray: string[];
  /**
   * Event 2 Property - Boolean
   */
  requiredBoolean: boolean;
  /**
   * Event 2 Property - Enum
   *
   * | Rule | Value |
   * |---|---|
   * | Enum Values | Enum1, Enum2 |
   */
  requiredEnum: "Enum1" | "Enum2";
  /**
   * Event 2 Property - Integer    *     * Examples:    * 5, 4, 3
   *
   * | Rule | Value |
   * |---|---|
   * | Type | integer |
   */
  requiredInteger: number;
  /**
   * Event 2 Property - Number
   *
   * | Rule | Value |
   * |---|---|
   * | Type | number |
   */
  requiredNumber: number;
  /**
   * Event 2 Property - String
   */
  requiredString: string;
}

export interface EventWithArrayTypesProperties {
  /**
   * description for required boolean array
   *
   * | Rule | Value |
   * |---|---|
   * | Item Type | boolean |
   */
  requiredBooleanArray: boolean[];
  /**
   * Description for required number array
   *
   * | Rule | Value |
   * |---|---|
   * | Item Type | number |
   */
  requiredNumberArray: number[];
  /**
   * Description for required object array
   */
  requiredObjectArray: any[];
  /**
   * description for required string array
   *
   * | Rule | Value |
   * |---|---|
   * | Item Type | string |
   */
  requiredStringArray: string[];
}

export interface EventWithDifferentCasingTypesProperties {
  /**
   * descriptionForEnumCamelCase
   *
   * | Rule | Value |
   * |---|---|
   * | Enum Values | enumCamelCase |
   */
  enumCamelCase: "enumCamelCase";
  /**
   * DescirptionForEnumPascalCase
   *
   * | Rule | Value |
   * |---|---|
   * | Enum Values | EnumPascalCase |
   */
  EnumPascalCase: "EnumPascalCase";
  /**
   * description_for_enum_snake_case
   *
   * | Rule | Value |
   * |---|---|
   * | Enum Values | enum_snake_case |
   */
  enum_snake_case: "enum_snake_case";
  /**
   * Description for enum with space
   *
   * | Rule | Value |
   * |---|---|
   * | Enum Values | enum with space |
   */
  "enum with space": "enum with space";
  /**
   * descriptionForCamelCase
   */
  propertyWithCamelCase: string;
  /**
   * DescriptionForPascalCase
   */
  PropertyWithPascalCase: string;
  /**
   * Description_for_snake_case
   */
  property_with_snake_case: string;
  /**
   * Description for case with space
   */
  "property with space": string;
}

export interface EventWithEnumTypesProperties {
  /**
   * Description for required enum
   *
   * | Rule | Value |
   * |---|---|
   * | Enum Values | optional enum 1, optional enum 2 |
   */
  "optional enum"?: "optional enum 1" | "optional enum 2";
  /**
   * Description for optional enum
   *
   * | Rule | Value |
   * |---|---|
   * | Enum Values | required enum 1, required enum 2 |
   */
  "required enum": "required enum 1" | "required enum 2";
}

export interface EventWithOptionalArrayTypesProperties {
  /**
   * Description for optional boolean array
   *
   * | Rule | Value |
   * |---|---|
   * | Item Type | boolean |
   */
  optionalBooleanArray?: boolean[];
  /**
   * Description for optional object array
   */
  optionalJSONArray?: any[];
  /**
   * Description for optional number array
   *
   * | Rule | Value |
   * |---|---|
   * | Item Type | number |
   */
  optionalNumberArray?: number[];
  /**
   * Description for optional string array
   *
   * | Rule | Value |
   * |---|---|
   * | Item Type | string |
   */
  optionalStringArray?: string[];
}

export interface EventWithOptionalPropertiesProperties {
  /**
   * | Rule | Value |
   * |---|---|
   * | Item Type | number |
   */
  optionalArrayNumber?: number[];
  /**
   * | Rule | Value |
   * |---|---|
   * | Item Type | string |
   */
  optionalArrayString?: string[];
  optionalBoolean?: boolean;
  /**
   * | Rule | Value |
   * |---|---|
   * | Type | number |
   */
  optionalNumber?: number;
  /**
   * Optional String property description
   */
  optionalString?: string;
}

export interface EventWithTemplatePropertiesProperties {
  /**
   * optional_event_property description
   *
   * | Rule | Value |
   * |---|---|
   * | Type | number |
   */
  optional_event_property?: number;
  /**
   * optional_template_property description
   *
   * | Rule | Value |
   * |---|---|
   * | Type | number |
   */
  optional_template_property?: number;
  /**
   * required_event_property description
   */
  required_event_property: string;
  /**
   * required_template_property description
   */
  required_template_property: string;
}

export interface EventTemplateProperties {
  /**
   * optional_template_property description
   *
   * | Rule | Value |
   * |---|---|
   * | Type | number |
   */
  optional_template_property?: number;
  /**
   * required_template_property description
   */
  required_template_property: string;
}

export interface SourceTemplateProperties {
  /**
   * description for context optionalEnum
   *
   * | Rule | Value |
   * |---|---|
   * | Enum Values | Value 1, Value 2 |
   */
  optionalEnum?: "Value 1" | "Value 2";
  /**
   * description for context requiredString
   */
  requiredString: string;
}

export class Identify implements Event {
  eventType = 'Identify';

  constructor(
    public eventProperties: IdentifyProperties,
  ) {}
}

export class Group implements Event {
  eventType = 'Group';

  constructor(
    public eventProperties: GroupProperties,
  ) {}
}

export class EventMaxIntForTest implements Event {
  eventType = 'EventMaxIntForTest';

  constructor(
    public eventProperties: EventMaxIntForTestProperties,
  ) {}
}

export class EventNoProperties implements Event {
  eventType = 'Event No Properties';
}

export class EventObjectTypes implements Event {
  eventType = 'Event Object Types';

  constructor(
    public eventProperties: EventObjectTypesProperties,
  ) {}
}

export class EventWithAllProperties implements Event {
  eventType = 'Event With All Properties';
  eventProperties: EventWithAllPropertiesProperties & {
    'requiredConst': "some-const-value";
  };

  constructor(
    eventProperties: EventWithAllPropertiesProperties,
  ) {
    this.eventProperties = {
      ...eventProperties,
      'requiredConst': "some-const-value",
    };
  }
}

export class EventWithArrayTypes implements Event {
  eventType = 'Event With Array Types';

  constructor(
    public eventProperties: EventWithArrayTypesProperties,
  ) {}
}

export class EventWithConstTypes implements Event {
  eventType = 'Event With Const Types';
  eventProperties = {
    'Boolean Const': true,
    'Integer Const': 10,
    'Number Const': 2.2,
    'String Const': "String-Constant",
    'String Const WIth Quotes': "\"String \"Const With\" Quotes\"",
    'String Int Const': 0,
  };
}

export class EventWithDifferentCasingTypes implements Event {
  eventType = 'event withDifferent_CasingTypes';

  constructor(
    public eventProperties: EventWithDifferentCasingTypesProperties,
  ) {}
}

export class EventWithEnumTypes implements Event {
  eventType = 'Event With Enum Types';

  constructor(
    public eventProperties: EventWithEnumTypesProperties,
  ) {}
}

export class EventWithOptionalArrayTypes implements Event {
  eventType = 'Event With Optional Array Types';

  constructor(
    public eventProperties?: EventWithOptionalArrayTypesProperties,
  ) {}
}

export class EventWithOptionalProperties implements Event {
  eventType = 'Event With Optional Properties';

  constructor(
    public eventProperties?: EventWithOptionalPropertiesProperties,
  ) {}
}

export class EventWithTemplateProperties implements Event {
  eventType = 'Event With Template Properties';

  constructor(
    public eventProperties: EventWithTemplatePropertiesProperties,
  ) {}
}

const getPromiseResponse = (result: boolean) => Promise.resolve(result);

// prettier-ignore
export class Ampli {
  private disabled: boolean = false;
  private amplitude?: Amplitude;

  get client(): Amplitude {
    this.isInitializedAndEnabled();
    return this.amplitude!;
  }

  private isInitializedAndEnabled(): boolean {
    if (!this.amplitude) {
      console.warn('WARNING: Ampli is not yet initialized. Have you called ampli.load() on app start?');
      return false;
    }
    return !this.disabled;
  }

  /**
   * Initialize the Ampli SDK. Call once when your application starts.
   * @param options Configuration options to initialize the Ampli SDK with.
   */
  load(options?: LoadOptions) {
    this.disabled = options?.disabled ?? false;

    if (this.amplitude) {
      console.warn('WARNING: Ampli is already intialized. Ampli.load() should be called once at application startup.');
      return { promise: getPromiseResponse(false) };
    }

    const env = options?.environment ?? 'development';
    const apiKey = options?.client?.apiKey ?? ApiKey[env];

    if (options?.client?.instance) {
      this.amplitude = options?.client?.instance;
      return { promise: getPromiseResponse(true) };
    }

    if (apiKey) {
      this.amplitude = Amplitude.getInstance();
      let promise = this.amplitude!.init(apiKey);

      const clientOptions = { ...DefaultOptions, ...options?.client?.options };
      const plan = clientOptions.plan;
      if (plan) {
        promise = promise.then(ok => ok ? this.amplitude!.setPlan(plan) : false);
      }
      return { promise };
    }

    console.warn("WARNING: ampli.load() requires 'environment', 'client.apiKey', or 'client.instance'");
    return { promise: getPromiseResponse(false) };
  }

  /**
   * Identify a user and set user properties.
   *
   * @param userId The user's id.
   * @param properties The user properties.
   * @param options Optional event options.
   * @param extra Extra unstructured data for middleware.
   */
  identify(
    userId: string | undefined,
    properties: IdentifyProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra
  ) {
    if (!this.isInitializedAndEnabled()) {
      return { promise: getPromiseResponse(false) };
    }

    this.handleEventOptions(options, userId);

    const amplitudeIdentify = new AmplitudeIdentify();
    if (properties != null) {
      for (const [key, value] of Object.entries(properties)) {
        amplitudeIdentify.set(key, value);
      }
    }

    const promise = this.amplitude!.identify(amplitudeIdentify, extra);
    return { promise };
  }

  setGroup(groupType: string, groupName: string | string[], options?: EventOptions, extra?: MiddlewareExtra) {
    if (!this.isInitializedAndEnabled()) {
      return { promise: getPromiseResponse(false) };
    }

    this.handleEventOptions(options);

    const promise = this.amplitude!.setGroup(groupType, groupName);
    return { promise };
  }

  /**
   * Identify a group and set group properties.
   *
   * @param groupType The group type.
   * @param groupName The group name.
   * @param properties The group properties.
   * @param options Optional event options.
   * @param extra Extra unstructured data for middleware.
   */
  groupIdentify(
    groupType: string,
    groupName: string | string[],
    properties: GroupProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    if (!this.isInitializedAndEnabled()) {
      return { promise: getPromiseResponse(false) };
    }

    this.handleEventOptions(options);

    const amplitudeIdentify = new AmplitudeIdentify();
    if (properties != null) {
      for (const [key, value] of Object.entries(properties)) {
        amplitudeIdentify.set(key, value);
      }
    }

    const promise = this.amplitude!.groupIdentify(groupType, groupName, amplitudeIdentify, extra);
    return { promise };
  }

  /**
   * Track event
   *
   * @param event The event to track.
   * @param options Optional event options.
   * @param extra Extra unstructured data for middleware.
   */
  track(event: Event, options?: EventOptions, extra?: MiddlewareExtra) {
    if (!this.isInitializedAndEnabled()) {
      return { promise: getPromiseResponse(false) };
    }

    this.handleEventOptions(options);

    const promise = this.amplitude!.logEvent(event.eventType, event.eventProperties, extra);
    return { promise };
  }

  flush() {
    if (!this.isInitializedAndEnabled()) {
      return { promise: getPromiseResponse(false) };
    }

    const promise = this.amplitude!.uploadEvents();
    return { promise };
  }

  /**
   * EventMaxIntForTest
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/EventMaxIntForTest)
   *
   * Event to test schema validation
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. intMax10)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventMaxIntForTest(
    properties: EventMaxIntForTestProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventMaxIntForTest(properties), options, extra);
  }

  /**
   * Event No Properties
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20No%20Properties)
   *
   * Event w no properties description
   *
   * Owner: Test codegen
   *
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventNoProperties(
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventNoProperties(), options, extra);
  }

  /**
   * Event Object Types
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20Object%20Types)
   *
   * Event with Object and Object Array
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. requiredObject)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventObjectTypes(
    properties: EventObjectTypesProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventObjectTypes(properties), options, extra);
  }

  /**
   * Event With All Properties
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20With%20All%20Properties)
   *
   * Event w all properties description
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. optionalString)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventWithAllProperties(
    properties: EventWithAllPropertiesProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventWithAllProperties(properties), options, extra);
  }

  /**
   * Event With Array Types
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20With%20Array%20Types)
   *
   * Description for event with Array Types
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. requiredBooleanArray)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventWithArrayTypes(
    properties: EventWithArrayTypesProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventWithArrayTypes(properties), options, extra);
  }

  /**
   * Event With Const Types
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20With%20Const%20Types)
   *
   * Description for event with const types
   *
   * Owner: Test codegen
   *
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventWithConstTypes(
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventWithConstTypes(), options, extra);
  }

  /**
   * event withDifferent_CasingTypes
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/event%20withDifferent_CasingTypes)
   *
   * Description for case with space
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. enumCamelCase)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventWithDifferentCasingTypes(
    properties: EventWithDifferentCasingTypesProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventWithDifferentCasingTypes(properties), options, extra);
  }

  /**
   * Event With Enum Types
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20With%20Enum%20Types)
   *
   * Description for event with enum types
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. optional enum)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventWithEnumTypes(
    properties: EventWithEnumTypesProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventWithEnumTypes(properties), options, extra);
  }

  /**
   * Event With Optional Array Types
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20With%20Optional%20Array%20Types)
   *
   * Description for event with optional array types
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. optionalBooleanArray)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventWithOptionalArrayTypes(
    properties?: EventWithOptionalArrayTypesProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventWithOptionalArrayTypes(properties), options, extra);
  }

  /**
   * Event With Optional Properties
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20With%20Optional%20Properties)
   *
   * Event w optional properties description
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. optionalArrayNumber)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventWithOptionalProperties(
    properties?: EventWithOptionalPropertiesProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventWithOptionalProperties(properties), options, extra);
  }

  /**
   * Event With Template Properties
   *
   * [View in Tracking Plan](https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20With%20Template%20Properties)
   *
   * Event with template properties description
   *
   * Owner: Test codegen
   *
   * @param properties The event's properties (e.g. optional_event_property)
   * @param options Amplitude event options.
   * @param extra Extra untyped parameters for use in middleware.
   */
  eventWithTemplateProperties(
    properties: EventWithTemplatePropertiesProperties,
    options?: EventOptions,
    extra?: MiddlewareExtra,
  ) {
    return this.track(new EventWithTemplateProperties(properties), options, extra);
  }

  private handleEventOptions(options: EventOptions | undefined, userId?: string) {
    if (userId || options?.userId) {
      this.amplitude?.setUserId(userId ?? options?.userId ?? null);
    }

    if (options?.deviceId) {
      this.amplitude?.setDeviceId(options.deviceId);
    }
  }
}

export const ampli = new Ampli();
