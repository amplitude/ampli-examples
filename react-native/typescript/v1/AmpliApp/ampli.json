{
  "OrgId": "132559",
  "WorkspaceId": "77b37977-cb3a-42eb-bce3-09f5f7c3adb7",
  "SourceId": "4afe79f0-c63a-484c-8589-11380377ead8",
  "Runtime": "react-native:typescript-ampli",
  "Path": "./src/ampli",
  "Branch": "main",
  "Version": "1.0.0",
  "VersionId": "a61c3908-ca4d-4c8d-8f81-54ad3ba17b9c",
  "OmitApiKeys": true,
  "Platform": "React Native",
  "Language": "TypeScript",
  "SDK": "@amplitude/react-native@^2.0"
}