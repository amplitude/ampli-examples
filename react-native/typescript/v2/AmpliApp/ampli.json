{
  "OrgId": "132559",
  "WorkspaceId": "77b37977-cb3a-42eb-bce3-09f5f7c3adb7",
  "SourceId": "86555606-49e6-427b-ab97-e5e60697f10b",
  "Runtime": "react-native:typescript-ampli-v2",
  "Path": "./src/ampli",
  "Branch": "main",
  "Version": "0.0.0",
  "VersionId": "79154a50-f057-4db5-9755-775e4e9f05e6",
  "OmitApiKeys": true,
  "Platform": "React Native",
  "Language": "TypeScript",
  "SDK": "@amplitude/analytics-react-native@^1.0"
}