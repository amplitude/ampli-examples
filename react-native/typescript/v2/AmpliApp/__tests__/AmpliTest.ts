import {Ampli, EventOptions, ReactNativeClient} from '../src/ampli';
import * as amplitude from '@amplitude/analytics-react-native';

type MockClient = {
  identify: jest.Mock;
  groupIdentify: jest.Mock;
  setGroup: jest.Mock;
  track: jest.Mock;
  flush: jest.Mock;
};

describe('ampli tests', () => {
  const userId = 'test-user-id';
  const options: EventOptions = {
    user_id: 'options-user-id',
    device_id: 'options-device-id',
  };

  const getVoidPromiseResult = () => ({promise: Promise.resolve()});

  let ampli: Ampli;
  const client: MockClient = {
    identify: jest.fn(getVoidPromiseResult),
    groupIdentify: jest.fn(getVoidPromiseResult),
    setGroup: jest.fn(getVoidPromiseResult),
    track: jest.fn(getVoidPromiseResult),
    flush: jest.fn(getVoidPromiseResult),
  };

  beforeEach(() => {
    ampli = new Ampli();
    ampli.load({client: {instance: client as unknown as ReactNativeClient}});
  });

  afterEach(() => {
    client.identify.mockClear();
    client.groupIdentify.mockClear();
    client.setGroup.mockClear();
    client.track.mockClear();
    client.flush.mockClear();
  });

  function checkSnapshots() {
    expect(client.identify.mock.calls).toMatchSnapshot('identify');
    expect(client.groupIdentify.mock.calls).toMatchSnapshot('groupIdentify');
    expect(client.setGroup.mock.calls).toMatchSnapshot('setGroup');
    expect(client.track.mock.calls).toMatchSnapshot('track');
    expect(client.flush.mock.calls).toMatchSnapshot('flush');
  }

  test('identify', async () => {
    await ampli.identify(
      userId,
      {
        requiredNumber: 123,
        optionalArray: ['A', 'BC'],
      },
      options,
    ).promise;
    checkSnapshots();
  });

  test('groupIdentify', async () => {
    const amplitudeIdentify = new amplitude.Identify();
    amplitudeIdentify.set('requiredBoolean', true);
    amplitudeIdentify.set('optionalString', 'abc');

    await ampli.client.groupIdentify(
      'test-group-type',
      'test-group-name',
      amplitudeIdentify,
      options,
    ).promise;
    checkSnapshots();
  });

  test('setGroup', async () => {
    await ampli.client.setGroup('test-group-type', 'test-group-name', options)
      .promise;
    checkSnapshots();
  });

  test('setGroup (multiple names)', async () => {
    await ampli.client.setGroup(
      'test-group-type',
      ['test-group-name1', 'test-group-name2'],
      options,
    ).promise;
    checkSnapshots();
  });

  test('eventNoProperties', async () => {
    await ampli.eventNoProperties(options).promise;
    checkSnapshots();
  });

  test('eventWithAllProperties', async () => {
    await ampli.eventWithAllProperties(
      {
        requiredArray: ['a', 'bc'],
        requiredBoolean: true,
        requiredEnum: 'Enum1',
        requiredInteger: 123,
        requiredNumber: 45.67,
        requiredString: 'required-string',
        optionalString: 'optional-string',
      },
      options,
    ).promise;
    checkSnapshots();
  });

  test('flush', async () => {
    await ampli.flush().promise;
    checkSnapshots();
  });
});
