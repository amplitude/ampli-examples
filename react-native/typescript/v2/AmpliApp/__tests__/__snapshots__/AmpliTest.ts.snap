// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ampli tests eventNoProperties: flush 1`] = `Array []`;

exports[`ampli tests eventNoProperties: groupIdentify 1`] = `Array []`;

exports[`ampli tests eventNoProperties: identify 1`] = `Array []`;

exports[`ampli tests eventNoProperties: setGroup 1`] = `Array []`;

exports[`ampli tests eventNoProperties: track 1`] = `
Array [
  Array [
    EventNoProperties {
      "event_type": "Event No Properties",
    },
    undefined,
    Object {
      "device_id": "options-device-id",
      "user_id": "options-user-id",
    },
  ],
]
`;

exports[`ampli tests eventWithAllProperties: flush 1`] = `Array []`;

exports[`ampli tests eventWithAllProperties: groupIdentify 1`] = `Array []`;

exports[`ampli tests eventWithAllProperties: identify 1`] = `Array []`;

exports[`ampli tests eventWithAllProperties: setGroup 1`] = `Array []`;

exports[`ampli tests eventWithAllProperties: track 1`] = `
Array [
  Array [
    EventWithAllProperties {
      "event_properties": Object {
        "optionalString": "optional-string",
        "requiredArray": Array [
          "a",
          "bc",
        ],
        "requiredBoolean": true,
        "requiredConst": "some-const-value",
        "requiredEnum": "Enum1",
        "requiredInteger": 123,
        "requiredNumber": 45.67,
        "requiredString": "required-string",
      },
      "event_type": "Event With All Properties",
    },
    undefined,
    Object {
      "device_id": "options-device-id",
      "user_id": "options-user-id",
    },
  ],
]
`;

exports[`ampli tests flush: flush 1`] = `
Array [
  Array [],
]
`;

exports[`ampli tests flush: groupIdentify 1`] = `Array []`;

exports[`ampli tests flush: identify 1`] = `Array []`;

exports[`ampli tests flush: setGroup 1`] = `Array []`;

exports[`ampli tests flush: track 1`] = `Array []`;

exports[`ampli tests groupIdentify: flush 1`] = `Array []`;

exports[`ampli tests groupIdentify: groupIdentify 1`] = `
Array [
  Array [
    "test-group-type",
    "test-group-name",
    Identify {
      "_properties": Object {
        "$set": Object {
          "optionalString": "abc",
          "requiredBoolean": true,
        },
      },
      "_propertySet": Set {
        "requiredBoolean",
        "optionalString",
      },
    },
    Object {
      "device_id": "options-device-id",
      "user_id": "options-user-id",
    },
  ],
]
`;

exports[`ampli tests groupIdentify: identify 1`] = `Array []`;

exports[`ampli tests groupIdentify: setGroup 1`] = `Array []`;

exports[`ampli tests groupIdentify: track 1`] = `Array []`;

exports[`ampli tests identify: flush 1`] = `Array []`;

exports[`ampli tests identify: groupIdentify 1`] = `Array []`;

exports[`ampli tests identify: identify 1`] = `
Array [
  Array [
    Identify {
      "_properties": Object {
        "$set": Object {
          "optionalArray": Array [
            "A",
            "BC",
          ],
          "requiredNumber": 123,
        },
      },
      "_propertySet": Set {
        "requiredNumber",
        "optionalArray",
      },
    },
    Object {
      "device_id": "options-device-id",
      "user_id": "test-user-id",
    },
  ],
]
`;

exports[`ampli tests identify: setGroup 1`] = `Array []`;

exports[`ampli tests identify: track 1`] = `Array []`;

exports[`ampli tests setGroup (multiple names): flush 1`] = `Array []`;

exports[`ampli tests setGroup (multiple names): groupIdentify 1`] = `Array []`;

exports[`ampli tests setGroup (multiple names): identify 1`] = `Array []`;

exports[`ampli tests setGroup (multiple names): setGroup 1`] = `
Array [
  Array [
    "test-group-type",
    Array [
      "test-group-name1",
      "test-group-name2",
    ],
    Object {
      "device_id": "options-device-id",
      "user_id": "options-user-id",
    },
  ],
]
`;

exports[`ampli tests setGroup (multiple names): track 1`] = `Array []`;

exports[`ampli tests setGroup: flush 1`] = `Array []`;

exports[`ampli tests setGroup: groupIdentify 1`] = `Array []`;

exports[`ampli tests setGroup: identify 1`] = `Array []`;

exports[`ampli tests setGroup: setGroup 1`] = `
Array [
  Array [
    "test-group-type",
    "test-group-name",
    Object {
      "device_id": "options-device-id",
      "user_id": "options-user-id",
    },
  ],
]
`;

exports[`ampli tests setGroup: track 1`] = `Array []`;
